src/chem/chem_system.cc

```diff
@@ -511,3 +531,554 @@ double ChemSystem::get_s2() const {
+                
+        double coef;
+        if (det2coef.count(new_det) == 1) coef=det2coef[new_det];
+        else continue;    
+
+        rdm(p,r) = rdm(p,r) + this_det.up.diff(new_det.up).permutation_factor*coef*coefs[idet];
+      } // r
+    } // i_elec
+
+    // dn electrons
+    for (unsigned i_elec = 0; i_elec < n_dn; i_elec++) {
+      unsigned p = occ_dn[i_elec];
+      for (unsigned r = 0; r < n_orbs; r++) {
+
+        if (orb_sym[p] != orb_sym[r]) continue;
+        if (p != r && this_det.dn.has(r)) continue;
+        
+        Det new_det = this_det;
+        new_det.dn.unset(p);
+        new_det.dn.set(r);
+        
+        double coef;
+        if (det2coef.count(new_det) == 1) coef=det2coef[new_det];
+        else continue;    
+      
+        rdm(p,r) = rdm(p,r) + this_det.dn.diff(new_det.dn).permutation_factor*coef*coefs[idet];
+      } // r
+    } // i_elec
+  } // idet
+//  std::cout<<rdm<<"\n";  
+  return rdm;
+//std::exit(0);  
+} 
+    
+    
+
+//======================================================
+void ChemSystem::generate_natorb_integrals(MatrixXd rdm) {
+// Compute natural orbitals by diagonalizing the 1RDM.
+// Rotate integrals to natural orbital basis and generate
+// new FCIDUMP file.
+// The original 1RDM (rdm) is left untouched.
+//======================================================
+  unsigned n_group_elements;
+  if (point_group == PointGroup::D2h) n_group_elements=8; 
+  
+  // Diagonalize rdm in the subspace of each irrep separately
+  std::vector<std::vector<unsigned>> inds(n_group_elements);
+  std::vector<unsigned> n_in_group(n_group_elements);
+
+  for (unsigned irrep = 0; irrep < n_group_elements; irrep++) {
+    n_in_group[irrep] = 0;
+    for (unsigned i=0; i<n_orbs; i++) {
+      if (orb_sym[i]==irrep+1) {
+        n_in_group[irrep] = n_in_group[irrep] + 1;
+        inds[irrep].push_back(i);
+      }
+    }
+  }
+  
+  double eigenvalues[n_orbs];
+  MatrixXd rot = MatrixXd::Zero(n_orbs,n_orbs); // rotation matrix
+
+  for (unsigned irrep=0; irrep<n_group_elements; irrep++) {
+    if (n_in_group[irrep] == 0) continue;
+    unsigned n = n_in_group[irrep];
+
+    MatrixXd tmp_rdm(n,n); // rdm in the subspace of current irrep
+    for (unsigned i=0; i<n; i++) {
+      for (unsigned j=0; j<n; j++) {
+        tmp_rdm(i,j) = rdm(inds[irrep][i],inds[irrep][j]);
+      }
+    }
+    
+    SelfAdjointEigenSolver<MatrixXd> es(tmp_rdm);   
+    MatrixXd tmp_eigenvalues, tmp_eigenvectors;
+    tmp_eigenvalues = es.eigenvalues().transpose();  
+    tmp_eigenvectors = es.eigenvectors();
+    
+    // columns of rot (rotation matrix) = eigenvectors of rdm
+    for (unsigned i=0; i<n; i++) {
+      eigenvalues[inds[irrep][i]] = tmp_eigenvalues(n-i-1);
+      for (unsigned j=0; j<n; j++) {
+        rot(inds[irrep][i],inds[irrep][j]) = tmp_eigenvectors(i,n-j-1); 
+      }
+    } 
+  } // irrep
+
+  std::cout<< "Occupation numbers:\n";
+  for (unsigned i=0; i<n_orbs;i++) std::cout<< eigenvalues[i] <<"\n";
```

Comment:
Could you format them as the other parts of the code? The format is specified by https://github.com/jl2922/shci/blob/master/.clang-format and works with all major code editors. On vim, I use https://github.com/google/vim-codefmt